src/java/org/apache/cassandra/io/sstable/SSTableZeroCopyWriter.java

```diff
@@ -61,12 +58,8 @@ public SSTableZeroCopyWriter(Builder<?, ?> builder,
         lifecycleNewTracker.trackNew(this);
         this.componentWriters = new HashMap<>();
 
-        if (!descriptor.getFormat().streamingComponents().containsAll(components))
-            throw new AssertionError(format("Unsupported streaming component detected %s",
-                                            Sets.difference(ImmutableSet.copyOf(components), descriptor.getFormat().streamingComponents())));
-
```

Comment:
@pkolaczk @mike-tr-adamson I'm not sure we want to just remove this validation, but checking SAI components raises some other questions. The simplest way around this for now might be...

```
Set<Component> unsupported = components.stream()
                                       .filter(c -> !descriptor.getFormat().streamingComponents().contains(c))
                                       .filter(c-> c.type != SSTableFormat.Components.Types.CUSTOM)
                                       .collect(Collectors.toSet());

if (!unsupported.isEmpty())
    throw new AssertionError(format("Unsupported streaming component detected %s", unsupported));
```

However, it might also be worth looking at why we use `CUSTOM` for SAI components in the first place. Is now the time to simply create a new first-class type in `Types`?

```
public static final Component.Type SAI = Component.Type.create("SAI", "SAI\+.*.db", null);
```

(The regex could be stricter, but details...)

If we do this, we the `CUSTOM` in the check above becomes `SAI`, we replace `CUSTOM` w/ `SAI` in a few places in the SAI code, like `Version`, and things should just work, right?

Reply:
btw, I just did exactly this, and it works fine...at least `IndexStreamingTest` has no issues

Reply:
Why do we want to artificially restrict the set of components streamed in the place that can stream any component?
It's not like streaming supports only some kind of components and doesn't support some other types. It will copy anything we'll throw at it.

Reply:
As for this:
```
public static final Component.Type SAI = Component.Type.create("SAI", "SAI\\+.*.db", null);
```

This wouldn't be very consistent with what we use component types for now in the core - because now each core sstable component  has its own type. So why should we throw all different SAI component types into a single type? 

I think we need to take another look at how custom components are registered and simply allow modules to provide their own types. So SAI component types should be defined in SAI. We could also allow the component type to decide whether it should be streamed or not.

Reply:
> I think we need to take another look at how custom components are registered and simply allow modules to provide their own types. So SAI component types should be defined in SAI. We could also allow the component type to decide whether it should be streamed or not.

I think this would work. +1

Reply:
Fixed